src/engine/naive_engine.cc

```diff
@@ -55,20 +69,41 @@ class NaiveEngine final : public Engine {
     opr->opr_name = opr_name;
     return opr;
   }
+
   void DeleteOperator(OprHandle op) override {
     NaiveOpr *opr = op->Cast<NaiveOpr>();
     delete opr;
   }
+
   void Push(OprHandle op, Context exec_ctx, int priority = 0, bool profiling = false) override {
     NaiveOpr *opr = op->Cast<NaiveOpr>();
-    this->PushAsync(opr->fn,
-                    exec_ctx,
-                    opr->const_vars,
-                    opr->mutable_vars,
-                    opr->prop,
-                    priority,
-                    PROFILER_MESSAGE(opr->opr_name));
+    opr->profiling = profiling;
+    this->PushAsync([&](RunContext ctx, CallbackOnComplete on_complete) {
+#if MXNET_USE_PROFILER
+        if (opr->profiling) {
+          opr->opr_stat = Profiler::Get()->AddOprStat(exec_ctx.dev_type, exec_ctx.dev_id);
+          uint64_t id = std::hash<std::thread::id>()(std::this_thread::get_id());
+          opr->opr_stat->thread_id = id;
+          opr->opr_stat->opr_name  = std::string(opr->opr_name);
+          CallbackOnComplete callback = CreateCallback(NaiveEngine::OnComplete, opr);
+          // record operator start timestamp
+          SetOprStart(opr->opr_stat);
+          opr->fn(ctx, callback);
+        } else {
+          opr->fn(ctx, on_complete);
+        }
+#else
+        opr->fn(ctx, on_complete);
+#endif
+      },
+      exec_ctx,
+      opr->const_vars,
+      opr->mutable_vars,
+      opr->prop,
+      priority,
+      PROFILER_MESSAGE(opr->opr_name));
   }
+
   void PushAsync(AsyncFn exec_fun,
```

Comment:
@piiswrong @ZihengJiang  i'm a little confused by the interface of `PushAsync()`, in ThreadedEngine, `PushAsync()` will call `Push()` for work, but in NativeEngine, 
Just the opposite, `Push()` called `PushAsync()` into it.